File: day10.py
with open("data/input10.txt") as f:
    L=[[int(a) for a in list(x)] for x in f.read().strip().split("\n")]
H,W=len(L),len(L[0])

def score(x,y,seen):
    if seen!=None:
        if (x,y) in seen: return 0
        seen.add((x,y))
    if L[y][x]==9: return 1
    total=0
    for dx,dy in [(-1,0),(1,0),(0,-1),(0,1)]:
        if x+dx<0 or x+dx>=W or y+dy<0 or y+dy>=H: continue
        if L[y+dy][x+dx]==L[y][x]+1: total+=score(x+dx,y+dy,seen)
    return total  

part1,part2=0,0
for y in range(H):
    for x in range(W):
        if L[y][x]==0:
            part1+=score(x,y,set()); part2+=score(x,y,None)
print(part1,part2)

File: day10.lua
local part1,part2=0,0
grid={}
for x in io.lines("data/input10.txt") do
    if #x>0 then
        local row={}
        for i=1,#x do
            table.insert(row,tonumber(string.sub(x,i,i)))
        end
        table.insert(grid,row)
    end
end
W,H=#grid,#grid[1]

function inside(x,y) return x>=1 and x<=W and y>=1 and y<=H end

function grade(x,y,ispart1)
    if ispart1 then
        if seen[300*y+x] then return 0 end
        seen[300*y+x]=1
    end
    if grid[y][x]==9 then return 1 end
    local score=0
    if inside(x-1,y) and grid[y][x-1]==grid[y][x]+1 then score=score+grade(x-1,y,ispart1) end
    if inside(x+1,y) and grid[y][x+1]==grid[y][x]+1 then score=score+grade(x+1,y,ispart1) end
    if inside(x,y-1) and grid[y-1][x]==grid[y][x]+1 then score=score+grade(x,y-1,ispart1) end
    if inside(x,y+1) and grid[y+1][x]==grid[y][x]+1 then score=score+grade(x,y+1,ispart1) end
    return score
end

for y=1,H do
    for x=1,W do
        if grid[y][x]==0 then
            seen={};part1=part1+grade(x,y,true)
            part2=part2+grade(x,y,false)
        end
    end
end
print(part1,part2)